[16.0][FIX] partner_firstname: fix installation w/ missing names

mohs8421 opened the pull request:
This change fixes this bug: https://github.com/OCA/partner-contact/issues/1677

With that the module can be installed as only partners with names will be processed

simahawk commented:
@mohs8421 could you please add a test for this change?
When you are done, could you please:
1. squash your commits
2. rewrite the commit message according to our guidelines https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#commit-message

Thanks!

mohs8421 commented:
I tried to write a test for it, but it is not possible with the module in place to test this, as the module itself prevents that.
So when trying to create a test record to run `_install_partner_firstname()` afterwards, it already exits with the empty name error.
Yes, that's partly the point of this module, but it should not be that strict during installation.

For reference, here is the testcode:
```python
    def test_module_installation(self):
        model = self.env[self.model].with_context(**self.context)
        unnamed = model.create({'name': '', 'email': 'anon@ymous.org'})
        named = model.create({'name': 'firstname lastname', 'email': 'firstname@lastname.org'})
        model._install_partner_firstname()
        self.assertEqual(unnamed.name, '')
        self.assertEqual((named.firstname, named.lastname), ('firstname', 'lastname'))
```

yvaucher reviewed (approved):
LGTM, seems safer that way.

SirAionTech commented:
> I tried to write a test for it, but it is not possible with the module in place to test this, as the module itself prevents that.
So when trying to create a test record to run `_install_partner_firstname()` afterwards, it already exits with the empty name error.
Yes, that's partly the point of this module, but it should not be that strict during installation.
> 
> For reference, here is the testcode:
> 
> ```python
>     def test_module_installation(self):
>         model = self.env[self.model].with_context(**self.context)
>         unnamed = model.create({'name': '', 'email': 'anon@ymous.org'})
>         named = model.create({'name': 'firstname lastname', 'email': 'firstname@lastname.org'})
>         model._install_partner_firstname()
>         self.assertEqual(unnamed.name, '')
>         self.assertEqual((named.firstname, named.lastname), ('firstname', 'lastname'))
> ```

You could `mock` whatever method is preventing you from creating `unnamed`, just while creating it, so that it's like you are creating the partner while the module is not installed yet.
You can see an example in https://github.com/OCA/account-invoicing/blob/86b9060ce96d91251f6a03bdc8021176880d3586/partner_invoicing_mode/tests/test_invoice_mode.py#L112:
```py
with mock.patch.object(ResPartner, "_update_next_invoice_date") as mock_update:
    ...
```

CRogos reviewed (commented):
@mohs8421 I think your test is nearly working. I've added your test to the test_empty.py file and just added the contact **"type": "invoice"** because empty name is not allowed on all contact types. Than the tests run successfully.


```
class AddressCase(TransactionCase):

...

    def test_module_installation(self):
        """Create an invoice patner without name."""
        model = self.env["res.partner"]
        unnamed = model.create({'name': '', 'email': 'anon@ymous.org', "type": "invoice"})
        named = model.create({'name': 'firstname lastname', 'email': 'firstname@lastname.org'})
        model._install_partner_firstname()
        self.assertEqual(unnamed.name, '')
        self.assertEqual((named.firstname, named.lastname), ('firstname', 'lastname'))
        
```

If you add your tests, and squash all commits, I will also approve this PR.

mohs8421 committed:
[16.0][FIX] partner_firstname: fix installation w/ missing names

Resolve issue blocking partner_firstname from installation

This change fixes this bug: https://github.com/OCA/partner-contact/issues/1677

With that the module can be installed as only partners with names will be processed
